### ISSUE TITLE
refGene convertor

### ISSUE THREAD

**Issue by cboustred** [robinandeer/chanjo#126]
Hi Robin,

Do you know of anyone who has written a convertor which will work on the refGene table from the UCSC database already?

I've had a hack, and tried to use my bed file to build a chanjo database but I'm getting the following error:

```python
sqlalchemy.exc.IntegrityError: (IntegrityError) column id is not unique u'INSERT INTO superblock (id, secondary_id, contig, start, "end", strand) VALUES (?, ?, ?, ?, ?, ?)' ((u'MIR6084', None, u'chr1', 20633679, 20633788, u'+'), (u'AP4B1-AS1', None, u'chr1', 113812612, 113901237, u'+'), (u'MIR6729', None, u'chr1', 12029158, 12029222, u'+'), (u'FLG-AS1', None, u'chr1', 152313459, 152366692, u'+'), (u'MIR6731', None, u'chr1', 24919345, 24919416, u'-'), (u'MIR6730', None, u'chr1', 12578957, 12579023, u'-'), (u'MIR6735', None, u'chr1', 43448539, 43448611, u'+'), (u'MIR4695', None, u'chr1', 18883202, 18883275, u'-')  ... displaying 10 of 31 total bound parameter sets ...  (u'NPPA-AS1', None, u'chr1', 11840319, 11847616, u'+'), (u'VIM-AS1', None, u'chr10', 17214239, 17229985, u'-'))
```

Assume because my hack at the refGene table was no good! I've tried to identify duplicates but can't see any obvious ones... Happy to share an example of my BED file

Basically I have a list of >6,000 gene names (all checked for HGNC). ~450 of these do not have a CCDS ID either because they have been withdrawn or because they are non-coding etc.

I thus want to look these 450 gene names up in the UCSC refGene table, convert to chanjo style BED format, combine with the CCDS chanjo bed and build a chanjo database.

Going to have another hack tomorrow, any tips much appreciated!

Thanks

Chris

**Comment by robinandeer** [comment]
Hi,

I'm afraid the only convert I'm aware of is the one that converts the CCDS database to Chanjo's BED format.

If you could share you BED file example and the file you are converting from that would help me help you a lot :+1:

**Comment by cboustred** [comment]
Hi Robin,

Thanks for getting back to me.

I've been around the houses a bit with this!

Long story short, I've made an updated BED file from querying HGNC gene ids
using data from the Ensembl REST API (attached).

I'm now able to make a Chanjo SQLite build with this file no problem,
however, I'm now trying the MySQL build and I'm getting the following error:

cat example.bed | chanjo --db chris:XXXXXXX@local_host/chanjo_test
--dialect "mysql" build
Traceback (most recent call last):
  File "/usr/local/bin/chanjo", line 9, in <module>
    load_entry_point('chanjo==2.3.0', 'console_scripts', 'chanjo')()
  File
"/usr/local/lib/python2.7/dist-packages/click-3.3-py2.7.egg/click/core.py",
line 610, in __call__
    return self.main(*args, **kwargs)
  File
"/usr/local/lib/python2.7/dist-packages/click-3.3-py2.7.egg/click/core.py",
line 590, in main
    rv = self.invoke(ctx)
  File
"/usr/local/lib/python2.7/dist-packages/click-3.3-py2.7.egg/click/core.py",
line 936, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File
"/usr/local/lib/python2.7/dist-packages/click-3.3-py2.7.egg/click/core.py",
line 782, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File
"/usr/local/lib/python2.7/dist-packages/click-3.3-py2.7.egg/click/core.py",
line 416, in invoke
    return callback(*args, **kwargs)
  File
"/usr/local/lib/python2.7/dist-packages/chanjo-2.3.0-py2.7.egg/chanjo/builder/cli.py",
line 28, in build
    overwrite=force
  File
"/usr/local/lib/python2.7/dist-packages/chanjo-2.3.0-py2.7.egg/chanjo/builder/core.py",
line 58, in init_db
    reduce(commit_per_contig(chanjo_db), superblocks, 'chr0')
  File
"/usr/local/lib/python2.7/dist-packages/toolz-0.7.1-py2.7.egg/toolz/functoolz.py",
line 216, in __call__
    return self._partial(*args, **kwargs)
  File
"/usr/local/lib/python2.7/dist-packages/chanjo-2.3.0-py2.7.egg/chanjo/builder/consumers.py",
line 29, in commit_per_contig
    chanjo_db.save()
  File
"/usr/local/lib/python2.7/dist-packages/chanjo-2.3.0-py2.7.egg/chanjo/store/core.py",
line 169, in save
    self.session.flush()
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/scoping.py",
line 150, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/session.py",
line 1919, in flush
    self._flush(objects)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/session.py",
line 2037, in _flush
    transaction.rollback(_capture_exception=True)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/util/langhelpers.py",
line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/session.py",
line 2001, in _flush
    flush_context.execute()
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/unitofwork.py",
line 372, in execute
    rec.execute(self)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/unitofwork.py",
line 526, in execute
    uow
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/persistence.py",
line 65, in save_obj
    mapper, table, insert)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/orm/persistence.py",
line 570, in _emit_insert_statements
    execute(statement, multiparams)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/engine/base.py",
line 729, in execute
    return meth(self, multiparams, params)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/sql/elements.py",
line 322, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/engine/base.py",
line 826, in _execute_clauseelement
    compiled_sql, distilled_params
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/engine/base.py",
line 958, in _execute_context
    context)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/engine/base.py",
line 1159, in _handle_dbapi_exception
    exc_info
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/util/compat.py",
line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/engine/base.py",
line 928, in _execute_context
    context)
  File
"/usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.9.8-py2.7-linux-x86_64.egg/sqlalchemy/connectors/mysqldb.py",
line 68, in do_executemany
    rowcount = cursor.executemany(statement, parameters)
  File "/usr/local/lib/python2.7/dist-packages/MySQLdb/cursors.py", line
262, in executemany
    r = self._query('\n'.join([query[:p], ',\n'.join(q), query[e:]]))
  File "/usr/local/lib/python2.7/dist-packages/MySQLdb/cursors.py", line
354, in _query
    rowcount = self._do_query(q)
  File "/usr/local/lib/python2.7/dist-packages/MySQLdb/cursors.py", line
318, in _do_query
    db.query(q)
sqlalchemy.exc.IntegrityError: (IntegrityError) (1452, 'Cannot add or
update a child row: a foreign key constraint fails (`chanjo_test`.`block`,
CONSTRAINT `block_ibfk_1` FOREIGN KEY (`superblock_id`) REFERENCES
`superblock` (`id`))') 'INSERT INTO block (id, secondary_id, contig, start,
end, strand, superblock_id) VALUES (%s, %s, %s, %s, %s, %s, %s)'
(('ENST00000207157', None, '1', 119425669, 119532179, '-', 'TBX15'),
('ENST00000497758', None, '1', 228295486, 228296274, '-', 'MRPL55'),
('ENST00000366903', None, '1', 221051699, 221058401, '+', 'HLX'),
('ENST00000482678', None, '1', 151395975, 151396939, '-', 'POGZ'),
('ENST00000295008', None, '1', 228294380, 228297000, '-', 'MRPL55'),
('ENST00000367137', None, '1', 205758221, 205782304, '-', 'SLC41A1'),
('ENST00000369836', None, '1', 110198703, 110204364, '+', 'GSTM4'),
('ENST00000468626', None, '1', 235992709, 236030176, '-', 'LYST')  ...
displaying 10 of 4106 total bound parameter sets ...  ('ENST00000369306',
None, '1', 145516252, 145523730, '+', 'PEX11B'), ('ENST00000310944', None,
'10', 88428426, 88459810, '+', 'LDB3'))

Odd it works in SQLite but not MySQL?

BW

Chris

**Comment by cboustred** [comment]
Hi Robin, sorry about this, I think the BED file attached to my last email was corrupted. Please ignore until I get my act together!

Chris
